Realm/ObjectServerTests/SwiftFlexibleSyncServerTests.swift

```diff
@@ -1315,6 +1356,337 @@ extension SwiftFlexibleSyncServerTests {
+                     "lastName": .string("M"),
+                     "age": .int32(30)
+                 ]
+        collection.insertOne(serverObject).await(self, timeout: 10.0)
+
+        let start = Date()
+        while collection.count(filter: [:]).await(self) != 11 && start.timeIntervalSinceNow > -10.0 {
+            sleep(1) // wait until server sync
+        }
+
+        // Resume the client session.
+        realm.syncSession!.resume()
+        XCTAssertEqual(results.count, 2)
+        results = try await realm.objects(SwiftPerson.self).where { $0.age >= 9 }.subscribe(waitForSync: .always)
+        // Expect this subscribe call to wait for sync downloads, even though the subscription already existed
+        XCTAssertEqual(results.count, 3) // Count is 3 because it includes the object/document that was created while offline.
+        XCTAssertEqual(realm.subscriptions.count, 1)
+    }
+
+    @MainActor
+    func testSubscribeNever() async throws {
+        try await populateSwiftPerson()
+        let realm = try openFlexibleSyncRealm()
+
+        let expectation = XCTestExpectation(description: "test doesn't hang")
+        Task {
+            let results = try await realm.objects(SwiftPerson.self).where { $0.age >= 8 }.subscribe(waitForSync: .never)
+            XCTAssertEqual(results.count, 0) // expect no objects to be able to sync because of immediate return
+            XCTAssertEqual(realm.subscriptions.count, 1)
+            expectation.fulfill()
+        }
+        await fulfillment(of: [expectation], timeout: 1)
+    }
+
+    @MainActor
+    func testSubscribeTimeout() async throws {
+        try await populateSwiftPerson()
+        let realm = try openFlexibleSyncRealm()
+
+        realm.syncSession!.suspend()
+        let expectation = XCTestExpectation(description: "doesn't wait longer than expected")
+        Task {
+            let timeout = 2.0
+            do {
+                _ = try await realm.objects(SwiftPerson.self).where { $0.age >= 8 }.subscribe(waitForSync: .always, timeout: timeout)
+            } catch let error as Realm.Error {
+                expectation.fulfill()
+                XCTAssertNotNil(error)
+                XCTAssertEqual(error.localizedDescription, "Waiting for update timed out after \(timeout) seconds.")
+                XCTAssertEqual(error.code, .clientTimeout)
+            }
+        }
+        await fulfillment(of: [expectation], timeout: 3.0)
+
+        // resume sync session and wait for subscription otherwise tear
+        // down can't complete successfuly
+        realm.syncSession!.resume()
+        let start = Date()
+        while realm.subscriptions.state != .complete && start.timeIntervalSinceNow > -10.0 {
+            sleep(1)
+        }
+        XCTAssertEqual(realm.subscriptions.state, .complete)
+    }
+
+    // fails in teardown "Assertion failed: no_sessions"
+    // can't tell which?
+    func skip_testSubscribeTimeoutSucceeds() async throws {
```

Comment:
Probably this is failing because you're not cancelling the dispatch block when completing succesfully, so this will probably going to fail when you set a timeout and successfully completes a subscription. Check https://developer.apple.com/documentation/dispatch/1431058-dispatch_block_cancel